src/control/lib/hardware/sysfs/provider.go

```diff
@@ -475,3 +475,17 @@ func condenseNetDevState(states []hardware.NetDevState) hardware.NetDevState {
 
 	return condensed
 }
+
+// IsIOMMUEnabled checks whether IOMMU is enabled by interrogating files in sysfs and implements
+// the IOMMUDetector interface on sysfs provider.
+func (s *Provider) IsIOMMUEnabled() (bool, error) {
+	if s == nil {
+		return false, errors.New("sysfs provider is nil")
+	}
```

Comment:
quibble: I know there was recent discussion about checking for nil receivers on struct methods, but I'm not sure I agree with this as a dogmatic rule. In this case, it seems to be pretty clearly a programming error to call this method on a nil `Provider`. In other cases like stringers and errors, it generally makes a more sense because sometimes nil pointers get passed into debug statements, etc. The only reason this method returns an error is to catch this specific programming error, so now we're forcing all callers to check for an error that indicates that they called the method on a nil receiver, and if that's a valid possibility, the caller should be checking that the thing isn't nil before trying to use it.

I would really prefer not to perpetuate this sort of thing throughout the codebase. I think defensive programming is generally a good idea, but in this case it feels a little overkill.

Reply:
I would rather return an informative error like this than just segfault. In the context of our CI at least it's easier to get an idea of what's messed up, even if it is a programming error. In recent experience it seems like our CI does not give us the Go coredumps.

Maybe it is a bit silly, but it doesn't hurt anything to check. It's a public method, so I think it makes sense to sanity check. If we don't want to return an error, and think this shouldn't ever happen, I think we should issue the panic ourselves.

Reply:
Given @mjmac is lead, I will follow his direction.